### PR TITLE
Pre-5.0 fixes

### PR DESCRIPTION
**Pull request by stgraber**
This does a bit more of DB cleanup and also adds a check for RSA certificates.

### CODE REVIEW

**Review by tomponline — APPROVED** [review]

File: lxd/certificates.go [hunk 1]
```diff
@@ -1039,3 +1034,26 @@ func certificateDelete(d *Daemon, r *http.Request) response.Response {
...
+		return fmt.Errorf("The provided certificate is expired")
+	}
+
+	if cert.PublicKeyAlgorithm == x509.ECDSA {
+		pubKey, ok := cert.PublicKey.(*rsa.PublicKey)
+		if !ok {
+			return fmt.Errorf("Unable to validate the RSA certificate")
+		}
+
+		if pubKey.Size() < 2048 {
```
Comment (simondeziel) [inline lxd/certificates.go:1053]: Looks like an `else if` branch was accidentally omitted.

**Review by simondeziel — CHANGES_REQUESTED** [review]

**Review by tomponline — APPROVED** [review]

File: lxd/certificates.go (hunk 1 above)
Comment (tomponline) [reply lxd/certificates.go:1053]: This seems fine as the !ok check will return so won't evaluate here.

Comment (simondeziel) [reply lxd/certificates.go:1053]: @tomponline isn't it inside `if cert.PublicKeyAlgorithm == x509.ECDSA {` ?

File: lxd/certificates.go [hunk 2]
```diff
@@ -1039,3 +1034,26 @@ func certificateDelete(d *Daemon, r *http.Request) response.Response {
...
+	if time.Now().Before(cert.NotBefore) {
+		return fmt.Errorf("The provided certificate isn't valid yet")
+	}
+
+	if time.Now().After(cert.NotAfter) {
+		return fmt.Errorf("The provided certificate is expired")
+	}
+
+	if cert.PublicKeyAlgorithm == x509.ECDSA {
+		pubKey, ok := cert.PublicKey.(*rsa.PublicKey)
```
Comment (simondeziel) [inline lxd/certificates.go:1048]: Maybe I'm wrong, but `x509.ECDSA` shouldn't have any `rsa.Publickey`, no?

File: lxd/certificates.go (hunk 1 above)
Comment (tomponline) [reply lxd/certificates.go:1053]: Hrm maybe it should be `cert.PublicKeyAlgorithm != x509.ECDSA`

There's no comments so can't tell what the intention was.

Comment (stgraber) [reply lxd/certificates.go:1053]: It should have been == x509.RSA, fixed now :)

File: lxd/certificates.go (hunk 2 above)
Comment (stgraber) [reply lxd/certificates.go:1048]: yeah, was meant to be == x509.RSA

**Review by tomponline — APPROVED** [review]
Third time lucky 😀

**Comment by stgraber** [comment]
The intent on the RSA check is to first check if the x509 parser tells us it's RSA, then if it tells us it is, not necessarily take it to its word that the struct is correct (because we'd panic if it's not) so validate that it is and fail with a pretty visible error if there's something dodgy going on.

This check may be able to catch someone trying to attack the x509 parser by confusing it about the type of key it's handling, maybe :)

**Pull request merged**